src/Mod/Sketcher/Gui/ViewProviderSketch.cpp

```diff
@@ -3481,6 +3519,218 @@ QIcon ViewProviderSketch::mergeColorfulOverlayIcons (const QIcon & orig) const
     return Gui::ViewProvider::mergeColorfulOverlayIcons (mergedicon);
 }
 
+bool ViewProviderSketch::getCameraMaxDimension(const Gui::View3DInventorViewer* viewer) {
+    float fHeight = -1.0;
+    float fWidth = -1.0;
+    viewer->getDimensions(fHeight, fWidth);
+    if (fHeight < 0 || fWidth < 0)
+        return false;
+
+    fHeight = std::max(fHeight, fWidth);
+    if (fabs(fHeight - camMaxDimension) > 0) { //ie if user zoomed.
+        camMaxDimension = fHeight;
+        return true;
+    }
+    return false;
+}
+
+bool ViewProviderSketch::getCameraCenter(const Gui::View3DInventorViewer* viewer)
+{
+    SoCamera* cam = viewer->getSoRenderManager()->getCamera();
+    if(!cam)
+        return false;
+
+    SbVec3f direction;
+    cam->orientation.getValue().multVec(SbVec3f(0, 0, -1), direction);
+    SbVec3f newCamCenterOnSketch = cam->position.getValue() + cam->focalDistance.getValue() * direction;
+
+    if ((camCenterOnSketch - newCamCenterOnSketch).length() > 0.1 * camMaxDimension) { //ie if user moved by more than 10% of the current 3d view dimensions.
+        camCenterOnSketch = newCamCenterOnSketch;
+        return true;
+    }
+    return false;
+}
+
+double ViewProviderSketch::getGridSize(const Gui::View3DInventorViewer* viewer)
+{
+    short pixelWidth = -1;
+    short pixelHeight = -1;
+    viewer->getViewportRegion().getViewportSizePixels().getValue(pixelWidth, pixelHeight);
+    if (pixelWidth < 0 || pixelHeight < 0)
+        return GridSize.getValue();
+
+    ParameterGrp::handle hGrp = App::GetApplication().GetParameterGroupByPath("User parameter:BaseApp/Preferences/Mod/Sketcher/General");
+    int numberOfLines = std::max(pixelWidth, pixelHeight) / hGrp->GetInt("GridSizePixelThreshold", 15);
+
+    double newGridSize = pow(hGrp->GetInt("GridNumberSubdivision", 10), 1 + floor(log10(camMaxDimension / numberOfLines)));

```

Comment:
IMO this should take into account the chosen user unit, folks using inches probably want their grid to be sane multiples of an inch
![image](https://user-images.githubusercontent.com/36372335/202190710-b389d376-d191-4fed-92df-a6823197ce88.png)

Reply:
also is pow on GridNumberSubdivision really the best scaling method? It works well for 10 but sounds like it would be weird with other values

Reply:
first I used pow on 10 but it wasn't working if number of subdivision wasn't 10. Pow on GridNumberSubdivision is the way it should to work on all subdivision numbers

Reply:
I agree with imperial units being insane, note that I am not asking for variable subdivisions based on scale, but at the very least decimal inch based grid should be supported. Hardcoding metric is not user friendly, sometimes you have to work with inches even if you hate them. I suggest if an inch based unit system is active `GridNumberSubdivision` should be mutiplied by 1in or something like that. In that case `numberOfLines` would probably also need to be adjusted to account for this.

About pow, I now tested other values for GridNumberSubdivision and agree that it works well.

Reply:
GridNumberSubdivision is an int. We could set GridNumberSubdivision to a value depending on user settings. Say defaut gridsize is 1in, then GridNumberSubdivision could be set to 12 so that big division would measure 1 foot.
But the issue is that if you zoom in and it resize to smaller grid. Then the division lines are 1 inch, and the smaller lines are 1/12th of inch. Which is meaningless.
So overall I think imperial units would need variable GridNumberSubdivision. Or the users would be better off using fixed grid size and properly set GridNumberSubdivision.

Reply:
> Imperial units are not sane by nature. They use different division at each order of magniture. So the number of subdivision has to be variable depending on grid size value. Which sounds like a real headache to code.
> I'm not supporting imperial units in general so I won't code it.
> If an imperial unit user want to improve the grid to support that he can come after me.

Too much judgement here. We shall keep away from that. Being factual is enough.

Reply:
Hmm I thought that the file used inch as a base unit when the user preference is inch. But in fact no. 1.0 is still 1.0mm even if user has inch. So indeed the grid is wrong when user use inch.
So yes multiplying by 1in is better. Not sure how to do that.

Reply:
how to unmark this as resolved? I made a quick test and multiplying GridNumberSubdivision by an inch doesn't give the desired result, this:
```cpp
double newGridSize = 2.54 * pow(hGrp->GetInt("GridNumberSubdivision", 10), 1 + floor(log10(camMaxDimension / numberOfLines / 2.54)));
```
gave a good behavior
![inchTest](https://user-images.githubusercontent.com/36372335/202213022-910f6dbf-da95-45b5-b2e8-1d0d799125a5.gif)

Reply:
I unresolved, but actually I just finished making the change. Thanks for the confirmation of the formulae. I'll just check it works as expected and push.